sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala

```diff
@@ -0,0 +1,354 @@
+    }
+    var replaced = false
+    val updated: Seq[NamedExpression] = {
+      for {
+        i <- 0 until project.projectList.size
+        projectExpr = project.projectList(i)
+        defaultExpr = if (i < defaultExprs.size) defaultExprs(i) else Literal(null)
+        colName = if (i < colNames.size) colNames(i) else ""
+      } yield projectExpr match {
+        case Alias(u: UnresolvedAttribute, _) if isExplicitDefaultColumn(u) =>
+          replaced = true
+          Alias(defaultExpr, colName)()
+        case u: UnresolvedAttribute if isExplicitDefaultColumn(u) =>
+          replaced = true
+          Alias(defaultExpr, colName)()
+        case expr@_ if expr.find{isExplicitDefaultColumn}.isDefined =>
+          // Return a more descriptive error message if the user tries to nest the DEFAULT column
+          // reference inside some other expression, such as DEFAULT + 1 (this is not allowed).
+          throw new AnalysisException(DEFAULTS_IN_EXPRESSIONS_ERROR)
+        case _ => projectExpr
+      }
+    }
+    if (!replaced) return None
+    Some(project.copy(projectList = updated))
+  }
+
+  /**
+   * Looks up the schema for the table object of an INSERT INTO statement from the catalog.
+   *
+   * @return the schema of the indicated table, or None if the lookup found nothing.
+   */
+  private def getInsertTableSchema: Option[StructType] = {
+    assert(insert.isDefined)
+    val lookup = try {
+      val tableName = insert.get.table match {
+        case r: UnresolvedRelation => TableIdentifier(r.name)
+        case r: UnresolvedCatalogRelation => r.tableMeta.identifier
+        case _ => return None
+      }
+      catalog.lookupRelation(tableName)
+    } catch {
+      case _: NoSuchTableException => return None
+    }
+    lookup match {
+      case SubqueryAlias(_, r: UnresolvedCatalogRelation) => Some(r.tableMeta.schema)
+      case _ => None
+    }
+  }
+}
+
+/**
+ * This object contains fields to help process DEFAULT columns with the below rule of the same name.
+ */
+object ResolveDefaultColumns {
+  // This column metadata indicates the default value associated with a particular table column that
+  // is in effect at any given time. Its value begins at the time of the initial CREATE/REPLACE
+  // TABLE statement with DEFAULT column definition(s), if any. It then changes whenever an ALTER
+  // TABLE statement SETs the DEFAULT. The intent is for this "current default" to be used by
+  // UPDATE, INSERT and MERGE, which evaluate each default expression for each row.
+  val CURRENT_DEFAULT_COLUMN_METADATA_KEY = "CURRENT_DEFAULT"
+  // This column metadata represents the default value for all existing rows in a table after a
+  // column has been added. This value is determined at time of CREATE TABLE, REPLACE TABLE, or
+  // ALTER TABLE ADD COLUMN, and never changes thereafter. The intent is for this "exist default"
+  // to be used by any scan when the columns in the source row are incomplete.
```

Comment:
I don't quite understand this. What do you mean by "columns in the source row are incomplete"?